lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs

```diff
@@ -307,15 +316,488 @@ impl Client {
+        match minor_function {
+            MinorFunction::IRP_MN_QUERY_DIRECTORY => {
+                let rdp_req = ServerDriveQueryDirectoryRequest::decode(device_io_request, payload)?;
+                debug!("received RDP: {:?}", rdp_req);
+                let file_id = rdp_req.device_io_request.file_id;
+                // https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_main.c#L610
+                if let Some(dir) = self.file_cache.get(file_id) {
+                    if dir.fso.file_type != FileType::Directory {
+                        return Err(invalid_data_error("received an IRP_MN_QUERY_DIRECTORY request for a file rather than a directory"));
+                    }
+
+                    // On the initial query, we need to get the list of files in this directory from
+                    // the client by sending a TDP SharedDirectoryListRequest.
+                    if rdp_req.initial_query != 0 {
+                        // https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_file.c#L775
+                        // TODO(isaiah): I'm observing that sometimes rdp_req.path will not be precisely equal to dir.path. For example, we will
+                        // get a ServerDriveQueryDirectoryRequest where path == "\\*", whereas the corresponding entry in the file_cache will have
+                        // path == "\\". I'm not quite sure what to do with this yet, so just leaving this as a note to self.
+                        let path = dir.path.clone();
+
+                        // Ask the client for the list of files in this directory.
+                        (self.tdp_sd_list_request)(SharedDirectoryListRequest {
+                            completion_id: rdp_req.device_io_request.completion_id,
+                            directory_id: rdp_req.device_io_request.device_id,
+                            path,
+                        })?;
+
+                        // When we get the response for that list of files...
+                        self.pending_sd_list_resp_handlers.insert(
+                            rdp_req.device_io_request.completion_id,
+                            Box::new(
+                                move |cli: &mut Self,
+                                      res: SharedDirectoryListResponse|
+                                      -> RdpResult<Vec<Vec<u8>>> {
+                                    if res.err_code == TdpErrCode::Nil {
+                                        // If SharedDirectoryListRequest succeeded, move the
+                                        // list of FileSystemObjects that correspond to this directory's
+                                        // contents to its entry in the file cache.
+                                        if let Some(dir) = cli.file_cache.get_mut(file_id) {
+                                            dir.contents = res.fso_list;
```

Comment:
The nesting here and number of branches is pretty complicated. Do you see any simplifications?